### PR TITLE
Display analytics up to yesterday instead of today

### PR DESCRIPTION
**Pull request by metamoni**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Stop displaying admin analytics for current day, as the data is incomplete. Instead, show analytics for any given period up to yesterday

## QA Instructions, Screenshots, Recordings
To QA this:
- go to `/admin`
- verify that analytics are displayed for a 7-day period leading up to "Yesterday"
- verify that the correct number of total Posts, Comments, Reactions and New Members are displayed
- verify that the chart displays the correct number of Posts, Comments, Reactions and New Members for each day

_Chart data_

https://user-images.githubusercontent.com/22390758/161431648-4be13b58-3330-4984-aa1b-cb2da8774dfb.mov

_Update post `published_at` date to show chart does not display data from today_

https://user-images.githubusercontent.com/22390758/161431611-9b316c8c-b96f-408f-80cf-de758cdc47d0.mov

### UI accessibility concerns?
This is a backend change with no impact on accessibility

## Added/updated tests?
- [X] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?
An empty circle described by comedian Demetri Martin as "a pie chart about procrastination"

### CODE REVIEW

File: app/services/admin/charts_data.rb [hunk 1]
```diff
@@ -5,15 +5,15 @@ def initialize(length = 7)
...
+      period = @length.days.ago..1.day.ago
       previous_period = (@length * 2).days.ago..@length.days.ago
 
       grouped_posts = Article.where(published_at: period).group("DATE(published_at)").size
       grouped_comments = Comment.where(created_at: period).group("DATE(created_at)").size
       grouped_reactions = Reaction.where(created_at: period).group("DATE(created_at)").size
       grouped_users = User.where(registered_at: period).group("DATE(registered_at)").size
 
-      days_range = (@length - 1).downto(0)
+      days_range = (@length).downto(0)
```
Comment (metamoni) [inline app/services/admin/charts_data.rb:16]: Forgot to remove the `()` here. Will fix that when PR is ready for review.

File: app/views/admin/overview/_analytics.html.erb [hunk 2]
```diff
@@ -61,7 +61,7 @@
           data-values="<%= values.join(",") %>"></canvas>
         <footer class="flex justify-between items-center fs-xs color-secondary">
           <span><%= @labels[0] %></span>
-          <span>Today</span>
+          <span>Yesterday</span>
```
Comment (metamoni) [inline app/views/admin/overview/_analytics.html.erb:64]: Do we want to localize this?

File: spec/system/admin/admin_visits_overview_spec.rb [hunk 3]
```diff
@@ -1,14 +1,54 @@
 require "rails_helper"
 
 RSpec.describe "Admin visits overview page", type: :system, js: true do
+  include ActiveSupport::Testing::TimeHelpers
```
Comment (metamoni) [inline spec/system/admin/admin_visits_overview_spec.rb:4]: Can also include this globally with

```ruby
RSpec.configure do |config|
  config.include ActiveSupport::Testing::TimeHelpers
end
```

File: spec/system/admin/admin_visits_overview_spec.rb [hunk 4]
```diff
@@ -1,14 +1,54 @@
...
+    it "starts with yesterday" do
+      expect(page.body).to include "Yesterday"
+    end
+
+    it "starts a week before" do
+      expect(page.body).to include "Mar 24"
+    end
+
+    it "displays correct number of posts from past week" do
+      create(:article, published_at: Time.zone.local(2022, 3, 30))
```
Comment (metamoni) [inline spec/system/admin/admin_visits_overview_spec.rb:31]: I added these as variables at the top, as part of `describe "analytics"`, but for some reason that didn't work and it found 0 articles inside my `it` blocks. Any idea how to group these at the top, so I can avoid creating them inside each example?

File: app/views/admin/overview/_analytics.html.erb (hunk 2 above)
Comment (djuber) [reply app/views/admin/overview/_analytics.html.erb:64]: Great question! 

We're building out the solution in the public facing site first - while this is a good target for localization in the future, right now /admin/ pages aren't being localized.

File: spec/system/admin/admin_visits_overview_spec.rb (hunk 4 above)
Comment (djuber) [reply spec/system/admin/admin_visits_overview_spec.rb:31]: You may either want to add the `create` calls to the `before` block, or via a `let!` binding. Create calls outside of an it block (and outside of a before block) won't be run during test execution.

File: spec/system/admin/admin_visits_overview_spec.rb [hunk 5]
```diff
@@ -1,14 +1,54 @@
...
+      create(:article, published_at: Time.zone.local(2022, 3, 24))
+      create(:article, published_at: Time.zone.local(2022, 3, 10))
+      page.evaluate_script("window.location.reload()")
+
+      expect(page.body).to include "2</span> Posts"
+    end
+
+    it "does not display posts from previous week" do
+      create(:article, published_at: Time.zone.local(2022, 3, 10))
+      page.evaluate_script("window.location.reload()")
```
Comment (djuber) [inline spec/system/admin/admin_visits_overview_spec.rb:42]: Since the post counts for the timeframe are rendered on the backend, and present in the html response, these test cases would work fine in a request spec (like spec/requests/admin/overview_spec.rb) rather than needing to use headless chrome to load the page. That's faster than a js enabled system test, and possibly closes another edge case.

I think the best reason to avoid putting this in a system test is that the timezone randomization we use (every time you run the specs you're assigned a random time zone) may only affect the rails server (times are modified to the server's local time), but the headless chrome process may be running on your system time, in CI that's UTC, but my computer or yours may differ. 

When the system time of chrome, and the rails server time set randomly during test execution, are more than 12 hours apart there appears to be a failure. For example, I'm UTC-5, running this test with `ZONEBIE_TZ="Australia/Perth"` consistently fails because the dates in the browser change relative to the backend (dates are localized by the browser) - my time shows "Mar 25 to Yesterday" rather than "Mar 24 to Yesterday". 

It's more clear that the headless browser is not respecting the timezone change here if you set the date farther away from today 

```
  describe "analytics" do
    before do
      travel_to Time.zone.local(2022, 4, 30)
    end

    it "starts a week before" do
      expect(page.body).to include "Apr 24"
    end
```

This fails showing March dates (my current last week) in the rendered postsChart, while the response was served as though it were April.

https://gist.github.com/djuber/c80ce1de406677336010804fb63eb52e this is basically moving the tests out of the system test to test the html only.

Comment (metamoni) [reply spec/system/admin/admin_visits_overview_spec.rb:42]: That makes sense, I was wondering whether this was the best place for these specs. But I do think adding _only_ request specs might not fully cover everything, because we're relying on the `charts_data` working properly too. I'll add some more examples in there too, to make sure we've covered everything. Thanks for all your feedback, @djuber!

File: spec/system/admin/admin_visits_overview_spec.rb (hunk 4 above)
Comment (metamoni) [reply spec/system/admin/admin_visits_overview_spec.rb:31]: Totally forgot the `!` on my `let`s - it's why it wasn't working 🤦🏻‍♀️

**Review by jeremyf — APPROVED** [review]

**Review by Ridhwana — APPROVED** [review]
This looks good to me! Thank you @metamoni ✨
